Dockerfile

```diff
@@ -1,8 +1,48 @@
 # syntax-docker/dockerfile:1
-FROM python:alpine3.14
+
+FROM python:3.11-slim AS builder
+
+# This is neccesary step for psycopg2 to function in slim image
+RUN apt-get update && \
+    apt-get install -y libpq-dev gcc
+
+# Create virtual environment
+RUN python -m venv /opt/venv
+# Activate venv
+ENV PATH="/opt/venv/bin:$PATH"
+
+COPY requirements.txt .
+RUN pip install -r requirements.txt
+
+
+# Operational stage
+FROM python:3.11-slim
```

Comment:
The two-stage pattern is much less useful in python.  If you look at these two, you're saving yourself maybe not installing `gcc`, which is a tiny piece of software.

For compiled programs, the two stages are great because after you compile the binary you can get rid of all the install stuff.  But for interpreted languages, you still need all the stuff to interpret it all the time.

So it's not bad, it's just not helpful, at least not in this case.